src/backend/x11/mod.rs

```diff
@@ -0,0 +1,921 @@
+
+        let (resize_send, resize_recv) = mpsc::channel();
+
+        let backend = X11Backend {
+            log: logger,
+            source,
+            connection,
+            window,
+            key_counter: Arc::new(AtomicU32::new(0)),
+            depth,
+            visual_id,
+            screen_number,
+            resize: resize_send,
+        };
+
+        let surface = X11Surface::new(&backend, format, resize_recv)?;
+
+        Ok((backend, surface))
+    }
+
+    /// Returns the default screen number of the X server.
+    pub fn screen(&self) -> usize {
+        self.screen_number
+    }
+
+    /// Returns the underlying connection to the X server.
+    pub fn connection(&self) -> &RustConnection {
+        &*self.connection
+    }
+
+    /// Returns a handle to the X11 window this input backend handles inputs for.
+    pub fn window(&self) -> Window {
+        self.window.clone().into()
+    }
+}
+
+/// An X11 surface which uses GBM to allocate and present buffers.
+#[derive(Debug)]
+pub struct X11Surface {
+    connection: Weak<RustConnection>,
+    window: Window,
+    resize: Receiver<Size<u16, Logical>>,
+    device: gbm::Device<RawFd>,
+    format: DrmFourcc,
+    width: u16,
+    height: u16,
+    current: Dmabuf,
+    next: Dmabuf,
+}
+
+impl X11Surface {
+    fn new(
+        backend: &X11Backend,
+        format: DrmFourcc,
+        resize: Receiver<Size<u16, Logical>>,
+    ) -> Result<X11Surface, X11Error> {
+        let connection = &backend.connection;
+        let window = backend.window();
+
+        // Determine which drm-device the Display is using.
+        let screen = &connection.setup().roots[backend.screen()];
+        let dri3 = connection.dri3_open(screen.root, 0)?.reply()?;
+
+        let drm_device_fd = dri3.device_fd;
+        // Duplicate the drm_device_fd
+        let drm_device_fd: RawFd = fcntl::fcntl(
+            drm_device_fd.as_raw_fd(),
+            fcntl::FcntlArg::F_DUPFD_CLOEXEC(3), // Set to 3 so the fd cannot become stdin, stdout or stderr
+        )
+        .map_err(AllocateBuffersError::from)?;
+
+        let fd_flags =
+            nix::fcntl::fcntl(drm_device_fd.as_raw_fd(), nix::fcntl::F_GETFD).expect("Handle this error");
+        // No need to check if ret == 1 since nix handles that.
+
+        // Enable the close-on-exec flag.
+        nix::fcntl::fcntl(
+            drm_device_fd.as_raw_fd(),
+            nix::fcntl::F_SETFD(
+                nix::fcntl::FdFlag::from_bits_truncate(fd_flags) | nix::fcntl::FdFlag::FD_CLOEXEC,
+            ),
+        )
+        .map_err(AllocateBuffersError::from)?;
+
+        if get_drm_node_type_from_fd(drm_device_fd.as_raw_fd())? != DRM_NODE_RENDER {
+            todo!("Attempt to get the render device by name for the DRM node that isn't a render node")
+        }
+
+        // Finally create a GBMDevice to manage the buffers.
+        let device = gbm::Device::new(drm_device_fd.as_raw_fd()).expect("Failed to create gbm device");
```

Comment:
I think this `as_raw_fd()` call doesn't do anything. At least this calls `RawFd::as_raw_fd()`.

...which makes me wonder: Who closes the FD? Looking at the code in `gbm::Device::new`, I think perhaps you want to keep x11rb's `RawFdContainer`. You can still set the `CLOEXEC` frag via `as_raw_fd()` on it. Since it implements `AsRawFd`, you should be able to just pass it as-is to `gbm::Device::new`. Only downside is that this would change the type of `device` from `Device<RawFd>` to `Device<RawFdContainer>`. But at least that would make sure the FD is actually closed when needed.

(Sorry if I misunderstood this gbm code and it actually closes the FD for you. It just really does not look like it...)

Reply:
Changing to `RawFdContainer` seems to cause a segfault when creating the device for some reason.